orchard_note_encryption.py

```diff
@@ -0,0 +1,295 @@
+
+        leadbyte = p_enc[0]
+        assert(leadbyte == 2)
+        np = OrchardNotePlaintext(
+            p_enc[1:12],   # d
+            Scalar.from_bytes(p_enc[12:20]),  # v
+            p_enc[20:52],  # rseed
+            p_enc[52:564], # memo
+        )
+
+        g_d = diversify_hash(np.d)
+        pk_d = OrchardKeyAgreement.derive_public(ivk, g_d)
+        note = OrchardNote(np.d, pk_d, np.v, rho, np.rseed)
+
+        esk = OrchardKeyAgreement.esk(np.rseed, rho)
+        if OrchardKeyAgreement.derive_public(esk, g_d) != epk:
+            return None
+
+        cm = note.note_commitment()
+        if not cm:
+            return None
+        if cm != cm_star:
+            return None
+
+        return (note, np.memo)
+
+    def decrypt_using_fvk(self, fvk, rseed, rho, cv, cm_star):
+        ock = prf_ock_orchard(fvk.ovk, bytes(cv), bytes(cm_star.extract()), bytes(self.epk))
+        op = OrchardSym.decrypt(ock, self.c_out)
+        if not op:
+            return None
+
+        (pk_d_star, esk) = (op[0:32], op[32:64])
+        esk = Scalar.from_bytes(esk)
+        pk_d = Point.from_bytes(pk_d_star)
+        if bytes(pk_d) != pk_d_star:
+            return None
+        if OrchardKeyAgreement.esk(rseed, rho) != esk:
+            return None
+
+        shared_secret = OrchardKeyAgreement.agree(esk, pk_d)
+        k_enc = kdf_orchard(shared_secret, self.epk)
+        p_enc = OrchardSym.decrypt(k_enc, self.c_enc)
+        if not p_enc:
+            return None
+
+        leadbyte = p_enc[0]
+        assert(leadbyte == 2)
+        np = OrchardNotePlaintext(
+            p_enc[1:12],   # d
+            Scalar.from_bytes(p_enc[12:20]),  # v
+            p_enc[20:52],  # rseed
+            p_enc[52:564], # memo
+        )
+        g_d = diversify_hash(np.d)
+        note = OrchardNote(np.d, pk_d, np.v, rho, np.rseed)
+
+        cm = note.note_commitment()
+        if not cm:
+            return None
+        if cm != cm_star:
+            return None
+
+        if OrchardKeyAgreement.derive_public(esk, g_d) != self.epk:
+            return None
+
+        return (note, np.memo)
+
+def main():
+    args = render_args()
+
+    from random import Random
+    rng = Random(0xabad533d)
+    def randbytes(l):
+        ret = []
+        while len(ret) < l:
+            ret.append(rng.randrange(0, 256))
+        return bytes(ret)
+    rand = Rand(randbytes)
+
+    ne = OrchardNoteEncryption(rand)
+
+    test_vectors = []
+    for _ in range(0, 10):
+        sender_sk = SpendingKey(rand.b(32))
+        sender_fvk = FullViewingKey(sender_sk)
+
+        receiver_sk = SpendingKey(rand.b(32))
+        receiver_fvk = FullViewingKey(receiver_sk)
+        ivk = receiver_fvk.ivk()
+        d = receiver_fvk.default_d()
+        pk_d = receiver_fvk.default_pkd()
+        g_d = diversify_hash(d)
+
+        rseed = ne.rseed()
+        memo = rand.b(512)
+        np = OrchardNotePlaintext(
+            d,
+            Scalar(rand.u64() % (MAX_MONEY + 1)),
```

Comment:
Orchard note values are `u64`s; we should not be constraining to the Zcash `MAX_MONEY` here (which would be incorrect for e.g. ZSAs).
```suggestion
            rand.u64(),
```